### PR TITLE
[public-api] Add ListWorkspaces support

### PR DESCRIPTION
**Pull request by csweichel**
## Description
This PR adds support for the `ListWorkspaces` operation on the public-api.

As part of this change I had to introduce the missing `preparing` phase on the public API enum. I took the liberty and ordered the numbers in that enum correctly. I believe this is ok to this time 'round because the enum wasn't used anywhere, hence there cannot be consumers of that enum yet.

### Using `gpctl api workspaces list`
This requires a Gitpod API token with the correct scopes. Easiest way to get ahold of such a token is to 
1. start a workspace
2. Export and copy the Gitpod token for the workspace:
    ```bash
    curl --silent localhost:22999/_supervisor/v1/token/gitpod/cw-public-api-listws.preview.gitpod-dev.com/ | jq -r .token
    ```
4. port-forward the DB pod and connect with mysql
5. Update the function scopes 
    ```
    UPDATE d_b_gitpod_token SET scopes = 'function:*,resource:default';
    ```
6. Set the `GPCTL_PUBLICAPI_TOKEN` env var to the token value
    ```bash
    export GPCTL_PUBLICAPI_TOKEN=...
    ```

image

## Release Notes

```release-note
[public-api] Add `ListWorkspaces` support
```

## Werft options:

- [x] /werft with-preview

### CODE REVIEW

File: components/public-api-server/pkg/server/integration_test.go [hunk 1]
```diff
@@ -38,9 +38,6 @@ func TestPublicAPIServer_v1_WorkspaceService(t *testing.T) {
 
 	workspaceClient := v1.NewWorkspacesServiceClient(conn)
 
-	_, err = workspaceClient.ListWorkspaces(ctx, &v1.ListWorkspacesRequest{})
```
Comment (easyCZ) [inline components/public-api-server/pkg/server/integration_test.go:41]: Would you be able to add a specific integration test for ListWorkspaces now that it's implemented, please?

File: components/public-api-server/pkg/apiv1/workspace_test.go [hunk 2]
```diff
@@ -184,6 +188,122 @@ func TestWorkspaceService_GetOwnerToken(t *testing.T) {
...
+			}
+
+			if diff := cmp.Diff(test.Expectation, act, protocmp.Transform()); diff != "" {
+				t.Errorf("unexpected convertWorkspaceInfo (-want +got):\n%s", diff)
+			}
+		})
+	}
+}
+
+func FuzzConvertWorkspaceInfo(f *testing.F) {
```
Comment (easyCZ) [inline components/public-api-server/pkg/apiv1/workspace_test.go:287]: nice!

File: components/public-api-server/pkg/apiv1/workspace.go [hunk 3]
```diff
@@ -119,3 +124,148 @@ func bearerTokenFromContext(ctx context.Context) (string, error) {
...
+	})
+	if err != nil {
+		return nil, proxy.ConvertError(err)
+	}
+
+	res := make([]*v1.ListWorkspacesResponse_WorkspaceAndInstance, 0, len(serverResp))
+	for _, ws := range serverResp {
+		workspaceAndInstance, err := convertWorkspaceInfo(ws)
+		if err != nil {
+			// TODO(cw): not sure this is a good idea - a single invalid workspace can break the entire response
```
Comment (easyCZ) [inline components/public-api-server/pkg/apiv1/workspace.go:152]: I think this is better than silently skipping the entry. I'd keep it like this, except I'd wrap the error in more context, such as `fmt.Errorf("failed to convert workspace instance ID: %s into response: %w", instanceID, err)`

File: components/gitpod-protocol/go/gitpod-service.go [hunk 4]
```diff
@@ -1754,6 +1754,7 @@ type WorkspaceInstanceStatus struct {
 	Phase        string                       `json:"phase,omitempty"`
 	Repo         *WorkspaceInstanceRepoStatus `json:"repo,omitempty"`
 	Timeout      string                       `json:"timeout,omitempty"`
+	Version      int                          `json:"version,omitempty"`
```
Comment (easyCZ) [inline components/gitpod-protocol/go/gitpod-service.go:1757]: Should this be here?

Comment (csweichel) [reply components/gitpod-protocol/go/gitpod-service.go:1757]: Yes. This is the field introduced in #11194 - it just wasn't added to the Go struct definition yet.

File: components/public-api-server/pkg/server/integration_test.go (hunk 1 above)
Comment (csweichel) [reply components/public-api-server/pkg/server/integration_test.go:41]: Done - I switched over to the gomock-generated `GitpodAPIInterface` implementation in the process. This allows for more fine-grained tests (e.g. passing the page size as limit).

File: components/public-api-server/pkg/apiv1/workspace_test.go (hunk 2 above)
Comment (csweichel) [reply components/public-api-server/pkg/apiv1/workspace_test.go:423]: haven't found anything with this yet though :)

Comment (easyCZ) [reply components/public-api-server/pkg/apiv1/workspace_test.go:423]: Just a really good implementation :)

**Review by easyCZ — APPROVED** [review]

**Comment by easyCZ** [comment]
Nice, thanks for the changes!

**Comment by csweichel** [comment]
> Nice, thanks for the changes!

Thank you for the super fast review turnarounds

File: components/public-api-server/pkg/apiv1/workspace.go (hunk 3 above)
Comment (csweichel) [reply components/public-api-server/pkg/apiv1/workspace.go]: Didn't find anything in 43min
image

File: dev/gpctl/cmd/public-api-workspaces-ownertoken.go [hunk 5]
```diff
@@ -0,0 +1,44 @@
...
+		workspaceID := args[0]
+
+		conn, err := newPublicAPIConn()
+		if err != nil {
+			log.Log.WithError(err).Fatal()
+		}
+
+		service := v1.NewWorkspacesServiceClient(conn)
+
+		log.Log.Debugf("Retrieving workspace ID: %s", workspaceID)
```
Comment (sagor999) [inline dev/gpctl/cmd/public-api-workspaces-ownertoken.go:27]:
```suggestion
		log.Log.Debugf("Retrieving owner token for workspace ID: %s", workspaceID)
```

Comment (csweichel) [reply dev/gpctl/cmd/public-api-workspaces-ownertoken.go]: Good catch - fixed :)

**Review by akosyakov — APPROVED** [review]
lgtm

**Pull request merged**